llvm/lib/Transforms/IPO/FunctionAttrs.cpp

```diff
@@ -580,6 +582,205 @@ struct ArgumentUsesTracker : public CaptureTracker {
   const SCCNodeSet &SCCNodes;
 };
 
+struct ArgumentUse {
+  Use *U;
+  std::optional<int64_t> Offset;
+};
+
+// A struct of argument access info. "Unknown" accesses are the cases like
+// unrecognized instructions, instructions that have more than one use of
+// the argument, or volatile memory accesses. "Unknown" implies "IsClobber"
+// and an empty access range.
+// Write or Read accesses can be clobbers as well for example, a Load with
+// scalable type.
+struct ArgumentAccessInfo {
+  enum AccessType { Write, Read, Unknown };
+  AccessType ArgAccessType;
+  ConstantRangeList AccessRanges;
+  bool IsClobber = false;
+};
+
+struct UsesPerBlockInfo {
+  DenseMap<Instruction *, ArgumentAccessInfo> Insts;
```

Comment:
How large is Insts typically?

DenseMap jumps from 0 to 64: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-densemap-h

SmallDenseMap<..., 8> or something might help make it more gradual 0 -> 8 -> 64

Reply:
Good point! According to an internal benchmark (muppet), 90% have a <= 2 size. Changed to SmallDenseMap<..., 2>.

Similarly to `UsesPerBlock`, 90% have a <=8 size so changed `UsesPerBlock` to SmallDenseMap<..., 8>

Reply:
If it's <= 2, and a non-trivial amount are == 2, you may want to provide 3 inline buckets to avoid having to grow() when the 2nd entry is inserted.
There's a max load factor of 3/4 https://github.com/llvm/llvm-project/blob/c1f1aab5d5d4d2a9d86e9e3e1794d609749a15f0/llvm/include/llvm/ADT/DenseMap.h#L612 so I think it won't allow filling 2 inline buckets with 2 entries.

Reply:
Oh, good point! Changed the SmallDenseMap size here to 4 and the size of `UsesPerBlock` to 16. 
With 4\*3/4 (<=3) and 16\*3/4 (<=12), they covers 96% and 94% cases in an internal large benchmark. Thanks!

BTW, the size must be a power of 2 so choose 4 and 16. https://github.com/llvm/llvm-project/blob/8f159096e097dde5a95d14984623c51d58ed8e69/llvm/include/llvm/ADT/DenseMap.h#L933